### PR TITLE
Add `--helm-debug` Flag to Kustomize for Enhanced Helm Debugging

### PR DESCRIPTION
**Pull request by isarns**
### **Summary of the Enhancement**
This PR adds a new flag, `--helm-debug`, to the Kustomize CLI. When this flag is provided, the `--debug` flag is passed to the Helm executable. This enables more detailed logging and error information from Helm, which is helpful when diagnosing issues with Helm charts during the Kustomize build process.

### **Why This is Useful**
In cases where Helm chart templates fail to render correctly, the default error messages can be opaque and difficult to diagnose. By enabling Helm’s `--debug` flag via `--helm-debug`, users can see detailed logs from Helm, including the specific files and lines where errors occur, as well as the internal processes Helm is performing. This can save significant time in troubleshooting and resolving issues.

### **Example Usage**
Without `--helm-debug`:
```sh
kustomize build . --load-restrictor=LoadRestrictionsNone --enable-helm
```

Output:
```
Error: Error: YAML parse error on s3/templates/bucket.yaml: error converting YAML to JSON: yaml: line 9: found character that cannot start any token

Use --debug flag to render out invalid YAML
: unable to run: 'helm template s3...'
```

With `--helm-debug`:
```sh
kustomize build . --load-restrictor=LoadRestrictionsNone --enable-helm --helm-debug
```

Output:
```
Error: install.go:218: [debug] Original chart version: ""
install.go:235: [debug] CHART PATH: /Users/isarn/git/IAC/crossplane/devops/live/devops/prod/ca-central-1/s3/iaas-scripts/charts/s3-0.1.1/s3

Error: YAML parse error on s3/templates/bucket.yaml: error converting YAML to JSON: yaml: line 9: found character that cannot start any token
helm.go:84: [debug] error converting YAML to JSON: yaml: line 9: found character that cannot start any token
YAML parse error on s3/templates/bucket.yaml
helm.sh/helm/v3/pkg/releaseutil.(*manifestFile).sort
        helm.sh/helm/v3/pkg/releaseutil/manifest_sorter.go:146
helm.sh/helm/v3/pkg/releaseutil.SortManifests
        helm.sh/helm/v3/pkg/releaseutil/manifest_sorter.go:106
helm.sh/helm/v3/pkg/action.(*Configuration).renderResources
        helm.sh/helm/v3/pkg/action/action.go:168
helm.sh/helm/v3/pkg/action.(*Install).RunWithContext
        helm.sh/helm/v3/pkg/action/install.go:304
main.runInstall
        helm.sh/helm/v3/cmd/helm/install.go:310
main.newTemplateCmd.func2
        helm.sh/helm/v3/cmd/helm/template.go:95
github.com/spf13/cobra.(*Command).execute
        github.com/spf13/cobra@v1.8.0/command.go:983
github.com/spf13/cobra.(*Command).ExecuteC
        github.com/spf13/cobra@v1.8.0/command.go:1115
github.com/spf13/cobra.(*Command).Execute
        github.com/spf13/cobra@v1.8.0/command.go:1039
main.main
        helm.sh/helm/v3/cmd/helm/helm.go:83
runtime.main
        runtime/proc.go:271
runtime.goexit
        runtime/asm_amd64.s:1695
---
# Source: s3/templates/bucket.yaml
apiVersion: s3.aws.upbound.io/v1beta1
kind: Bucket
metadata:
  labels:
    project: 
    env: 
    region: us-west-2
    crossplane-name: my-bucket
  name: %!!(MISSING)s(<nil>)-%!!(MISSING)s(<nil>)-usw2-my-bucket
spec:
  forProvider:
    tags:
      Name: %!!(MISSING)s(<nil>)-%!!(MISSING)s(<nil>)-usw2-my-bucket
  providerConfigRef:
    name:%!!(MISSING)s(<nil>)-%!!(MISSING)s(<nil>)
: unable to run: 'helm template s3 ...'
...
```

### **How to Test**
1. Add the `--helm-debug` flag to your Kustomize command as shown above.
2. Run the command against a Helm chart with known issues.
3. Verify that the detailed debug output is displayed, helping to identify the root cause of the error.

### **Additional Notes**
- This enhancement is backward compatible and does not interfere with the existing Helm integration in Kustomize.
- The `--helm-debug` flag is optional and can be omitted if detailed Helm debugging is not required.

### CODE REVIEW

File: api/internal/builtins/HelmChartInflationGenerator.go [hunk 1]
```diff
@@ -174,7 +177,7 @@ func (p *HelmChartInflationGeneratorPlugin) runHelmCommand(
 			fmt.Errorf(
 				"unable to run: '%s %s' with env=%s (is '%s' installed?): %w",
 				helm, strings.Join(args, " "), env, helm, err),
-			stderr.String(),
+				stderr.String()+stdout.String(),
```
Comment (stormqueen1990) [inline api/internal/builtins/HelmChartInflationGenerator.go:180]: I'm curious about the concatenation of the `stdout` content here. What's the scenario this aims to address?

File: api/types/helmchartargs.go
Comment (stormqueen1990) [inline api/types/helmchartargs.go:1]: Could you please add a test to `helmchartargs_test.go` to cover this new flag addition?

File: go.work.sum
Comment (stormqueen1990) [inline go.work.sum:1]: This change in `go.work.sum` seems unrelated to the code changes. Could you please revert it?

File: plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go
Comment (stormqueen1990) [inline plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go:1]: Can this change also be covered by a test?

**Review by stormqueen1990 — COMMENTED** [review]
Hello there, @isarns! 👋🏻

Thanks for your PR! I left a few comments, mostly around testing. Could you please take a look at those?

File: api/internal/builtins/HelmChartInflationGenerator.go (hunk 1 above)
Comment (isarns) [reply api/internal/builtins/HelmChartInflationGenerator.go:180]: `helm --debug` write both to stderr (stack trace) and to stdout for the incomplete YAML template.

File: plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go
Comment (isarns) [reply plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go:1]: I tried to write a test for that, but the flag only helps when a helm chart is broken, resulting in exit(1).

I used `rm := th.LoadAndRunGenerator` with a chart that returns error.
This will result in 
```
	rm, err := g.Generate()
	if err != nil {
		th.t.Fatalf("generate err: %v", err)
	}
```
Which fails the test. 
I tried to capture the stderr and stdout, but I didn't succeed in that. 

Please let me know if you have another idea for testing it.

File: api/internal/builtins/HelmChartInflationGenerator.go (hunk 1 above)
Comment (stormqueen1990) [reply api/internal/builtins/HelmChartInflationGenerator.go:180]: I see. Perhaps there would be value to adding a prefix so the user can identify which output is `stdout` and which output is `stderr`?
I am a bit concerned that this might become a wall of text difficult to parse through.

File: plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go
Comment (stormqueen1990) [reply plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go:1]: Just checking that the options are being passed through is enough. I agree that having a full end-to-end test for this case is difficult and taxing, and I was mainly aiming for verification that Kustomize is passing the options it claims to pass.

Comment (isarns) [reply plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go:1]: I added TestHelmChartInflationGeneratorWithDebug 
with `debug: true` the test is the same as TestHelmChartInflationGeneratorUsingVersionWithoutRepo
Because when `--debug` is present, but there were no issues in the helm template the flag does nothing.

Comment (isarns) [reply plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go:1]: I made a manual test

```yaml
# kustomization.yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

generators:
  - helmgenerator.yaml
```

```yaml
# helmgenerator.yaml
apiVersion: builtin
kind: HelmChartInflationGenerator
metadata:
  name: helm-debug
name: helm-debug
version: 1.0.0
releaseName: helm-debug-test
chartHome: ./
debug: true
valuesInline:
  namePrefix: ""
  name: ""
```

The output with kustomize v5.3.0
```
Error: Error: YAML parse error on helm-debug/templates/test.yaml: error converting YAML to JSON: yaml: line 4: block sequence entries are not allowed in this context

Use --debug flag to render out invalid YAML
: unable to run: 'helm template helm-debug-test /Users/isarn/temp/helm-debug -f /var/folders/6d/ck0zlsjx3vsd4x9r42y6knvr0000gn/T/kustomize-helm-3602805025/helm-debug-kustomize-values.yaml' with env=[HELM_CONFIG_HOME=/var/folders/6d/ck0zlsjx3vsd4x9r42y6knvr0000gn/T/kustomize-helm-3602805025/helm HELM_CACHE_HOME=/var/folders/6d/ck0zlsjx3vsd4x9r42y6knvr0000gn/T/kustomize-helm-3602805025/helm/.cache HELM_DATA_HOME=/var/folders/6d/ck0zlsjx3vsd4x9r42y6knvr0000gn/T/kustomize-helm-3602805025/helm/.data] (is 'helm' installed?): exit status 1
```

And the output with the new kustomize is 
```
Error: install.go:218: [debug] Original chart version: ""
install.go:235: [debug] CHART PATH: /Users/isarn/temp/helm-debug

Error: YAML parse error on helm-debug/templates/test.yaml: error converting YAML to JSON: yaml: line 4: block sequence entries are not allowed in this context
helm.go:84: [debug] error converting YAML to JSON: yaml: line 4: block sequence entries are not allowed in this context
YAML parse error on helm-debug/templates/test.yaml
helm.sh/helm/v3/pkg/releaseutil.(*manifestFile).sort
        helm.sh/helm/v3/pkg/releaseutil/manifest_sorter.go:146
helm.sh/helm/v3/pkg/releaseutil.SortManifests
        helm.sh/helm/v3/pkg/releaseutil/manifest_sorter.go:106
helm.sh/helm/v3/pkg/action.(*Configuration).renderResources
        helm.sh/helm/v3/pkg/action/action.go:168
helm.sh/helm/v3/pkg/action.(*Install).RunWithContext
        helm.sh/helm/v3/pkg/action/install.go:304
main.runInstall
        helm.sh/helm/v3/cmd/helm/install.go:310
main.newTemplateCmd.func2
        helm.sh/helm/v3/cmd/helm/template.go:95
github.com/spf13/cobra.(*Command).execute
        github.com/spf13/cobra@v1.8.0/command.go:983
github.com/spf13/cobra.(*Command).ExecuteC
        github.com/spf13/cobra@v1.8.0/command.go:1115
github.com/spf13/cobra.(*Command).Execute
        github.com/spf13/cobra@v1.8.0/command.go:1039
main.main
        helm.sh/helm/v3/cmd/helm/helm.go:83
runtime.main
        runtime/proc.go:271
runtime.goexit
        runtime/asm_amd64.s:1695
---
# Source: helm-debug/templates/test.yaml
---
apiVersion: test.kustomize.io/v1
kind: ValuesMergeTest
metadata:
  name: -
: unable to run: 'helm template helm-debug-test /Users/isarn/temp/helm-debug -f /var/folders/6d/ck0zlsjx3vsd4x9r42y6knvr0000gn/T/kustomize-helm-718105538/helm-debug-kustomize-values.yaml --debug' with env=[HELM_CONFIG_HOME=/var/folders/6d/ck0zlsjx3vsd4x9r42y6knvr0000gn/T/kustomize-helm-718105538/helm HELM_CACHE_HOME=/var/folders/6d/ck0zlsjx3vsd4x9r42y6knvr0000gn/T/kustomize-helm-718105538/helm/.cache HELM_DATA_HOME=/var/folders/6d/ck0zlsjx3vsd4x9r42y6knvr0000gn/T/kustomize-helm-718105538/helm/.data] (is 'helm' installed?): exit status 1
```

The ` --debug` is added to the helm command.

File: api/internal/builtins/HelmChartInflationGenerator.go (hunk 1 above)
Comment (isarns) [reply api/internal/builtins/HelmChartInflationGenerator.go]: I added an error format. 

The output looks like that now: 
```
Error:  Helm stack trace:
install.go:218: [debug] Original chart version: ""
install.go:235: [debug] CHART PATH: /Users/isarn/temp/helm-debug

Error: YAML parse error on helm-debug/templates/test.yaml: error converting YAML to JSON: yaml: line 4: block sequence entries are not allowed in this context
helm.go:84: [debug] error converting YAML to JSON: yaml: line 4: block sequence entries are not allowed in this context
YAML parse error on helm-debug/templates/test.yaml
helm.sh/helm/v3/pkg/releaseutil.(*manifestFile).sort
        helm.sh/helm/v3/pkg/releaseutil/manifest_sorter.go:146
helm.sh/helm/v3/pkg/releaseutil.SortManifests
        helm.sh/helm/v3/pkg/releaseutil/manifest_sorter.go:106
helm.sh/helm/v3/pkg/action.(*Configuration).renderResources
        helm.sh/helm/v3/pkg/action/action.go:168
helm.sh/helm/v3/pkg/action.(*Install).RunWithContext
        helm.sh/helm/v3/pkg/action/install.go:304
main.runInstall
        helm.sh/helm/v3/cmd/helm/install.go:310
main.newTemplateCmd.func2
        helm.sh/helm/v3/cmd/helm/template.go:95
github.com/spf13/cobra.(*Command).execute
        github.com/spf13/cobra@v1.8.0/command.go:983
github.com/spf13/cobra.(*Command).ExecuteC
        github.com/spf13/cobra@v1.8.0/command.go:1115
github.com/spf13/cobra.(*Command).Execute
        github.com/spf13/cobra@v1.8.0/command.go:1039
main.main
        helm.sh/helm/v3/cmd/helm/helm.go:83
runtime.main
        runtime/proc.go:271
runtime.goexit
        runtime/asm_amd64.s:1695

Helm template:
---
# Source: helm-debug/templates/test.yaml
---
apiVersion: test.kustomize.io/v1
kind: ValuesMergeTest
metadata:
  name: -

: unable to run: 'helm template helm-debug-test /Users/isarn/temp/helm-debug -f /var/folders/6d/ck0zlsjx3vsd4x9r42y6knvr0000gn/T/kustomize-helm-194917191/helm-debug-kustomize-values.yaml --debug' with env=[HELM_CONFIG_HOME=/var/folders/6d/ck0zlsjx3vsd4x9r42y6knvr0000gn/T/kustomize-helm-194917191/helm HELM_CACHE_HOME=/var/folders/6d/ck0zlsjx3vsd4x9r42y6knvr0000gn/T/kustomize-helm-194917191/helm/.cache HELM_DATA_HOME=/var/folders/6d/ck0zlsjx3vsd4x9r42y6knvr0000gn/T/kustomize-helm-194917191/helm/.data] (is 'helm' installed?): exit status 1
```

**Comment by isarns** [comment]
Thanks @stormqueen1990!

- Added the test. I'm not quite sure if this is the best way to test it as I commented on your comment.
- Formatted the debug output `Helm stack trace:` for stderr (the stack trace) and `Helm template:` for stdout (the incomplete template)

Please let me know if there is anything else that needs to be done.

File: go.work.sum
Comment (stormqueen1990) [inline go.work.sum:1]: This could be reverted for this PR, since it doesn't appear like new libraries were added.

**Review by stormqueen1990 — COMMENTED** [review]
Thanks @isarns!

/lgtm
/assign @koba1t
for maintainer review

File: go.work.sum
Comment (isarns) [reply go.work.sum:1]: Thanks! Reverted that. 

How can I make sure that those changes are not necessary? so it wont happen again.

Comment (stormqueen1990) [reply go.work.sum:1]: I unfortunately don't have a good answer to this question 😅 but usually if there were no additions to `go.mod` an update to `go.work.sum` is probably not required.

**Review by stormqueen1990 — COMMENTED** [review]
/lgtm
